### ISSUE TITLE
[REST API] Erratic behavior of Content Rest API for Contents with Categories Fields

### ISSUE THREAD

**Issue by joseorsini** [dotCMS/core#12739]
On any REST API call, we display Category Names once contents are pulled through any REST API call, however, when it comes to saving, we only allow Category Keys/Variables, so the Names we originally got from first pulls are useless and they cannot be reused on new versions of contents created through REST API calls.

## Expected Behavior
Behavior of categories should be consistent and configurable at the same time. IF the system allows storing contents with category fields by using their [Keys, Variables or even Inodes](https://github.com/dotCMS/core/blob/master/dotCMS/src/main/java/com/dotcms/rest/ContentResource.java#L809-L841), once these contents are pulled, we should follow the same logic in order to re-use these values if needed for future edits via REST API call.

## Current Behavior
Any GET operation made to REST API Call will return an Array of Categories Names, however, these ones cannot be used for Updating the same content we pulled on the first place.

## Possible solution:
Set a Property to configure whether we should use Keys, Variables or Ids for Categories on any GET/PUT action related to REST API, and set a proper REST Endpoint that could return more info from a category, given its id/key/variable (which are unique per Category)

## Steps to Reproduce (for bugs)

1. Create a Test Content Type called "Test Content 002".
2. Add the following fields:
    - Title: Text-type, Required, indexed, User-Searchable.
    - Categories: Select "Topic" category.
3. Go to Categories tool. Under Topic Category, edit "Retirement" one and add "retirement as its key"

<img width="1527" alt="screen shot 2017-10-03 at 1 38 22 pm" src="https://user-images.githubusercontent.com/1717110/31146025-9a6daa38-a842-11e7-96ad-1ffe7dd5531d.png">

4. Via curl, create a content using "Research", "Retirement" and "Private Banking" as Categories.

```
$ curl -v -u "admin@dotcms.com:admin" -XPUT http://localhost:8080/api/content/publish/1 -F 'json={"stName":"TestContent002","title":"Test 001 - Created through curl","categories":"Research, Retirement, Private Banking"};type=application/json'
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
* Server auth using Basic with user 'admin@dotcms.com'
> PUT /api/content/publish/1 HTTP/1.1
> Host: localhost:8080
> Authorization: Basic YWRtaW5AZG90Y21zLmNvbTphZG1pbg==
> User-Agent: curl/7.54.0
> Accept: */*
> Content-Length: 293
> Expect: 100-continue
> Content-Type: multipart/form-data; boundary=------------------------09175e4ddacb75ee
> 
< HTTP/1.1 100 Continue
< HTTP/1.1 200 OK
< Server: Apache-Coyote/1.1
< Set-Cookie: JSESSIONID=AD8177A119333863179AB12D70CAB895; Path=/; HttpOnly
< Location: http://localhost:8080/api/content/inode/329672e7-d9bc-4b2b-b37d-e4b512a1e807
< inode: 329672e7-d9bc-4b2b-b37d-e4b512a1e807
< identifier: f48d4c83-98aa-4157-9743-bb3936d15a09
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET, HEAD, POST, PUT, DELETE, OPTIONS
< Access-Control-Allow-Headers: Authorization, Accept, Content-Type, Cookies
< Content-Length: 0
< Date: Tue, 03 Oct 2017 19:49:19 GMT
< 
* Connection #0 to host localhost left intact
```
5. Pull the content you just created:

```
$ $ curl -v -u admin@dotcms.com:admin -XGET http://localhost:8080/api/content/query/+contentType:TestContent002
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
* Server auth using Basic with user 'admin@dotcms.com'
> GET /api/content/query/+contentType:TestContent002 HTTP/1.1
> Host: localhost:8080
> Authorization: Basic YWRtaW5AZG90Y21zLmNvbTphZG1pbg==
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: Apache-Coyote/1.1
< Set-Cookie: JSESSIONID=5D6A0AFA10883F7662D99CC59A48F294; Path=/; HttpOnly
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET, HEAD, POST, PUT, DELETE, OPTIONS
< Access-Control-Allow-Headers: Authorization, Accept, Content-Type, Cookies
< Content-Type: text/plain
< Content-Length: 402
< Date: Tue, 03 Oct 2017 19:50:34 GMT
< 
* Connection #0 to host localhost left intact
{"contentlets":[{"owner":"dotcms.org.1","identifier":"f48d4c83-98aa-4157-9743-bb3936d15a09","modDate":"2017-10-03 13:49:19.0","languageId":1,"title":"Test 001 - Created through curl","inode":"329672e7-d9bc-4b2b-b37d-e4b512a1e807","folder":"SYSTEM_FOLDER","sortOrder":0,"modUser":"dotcms.org.1","host":"SYSTEM_HOST","categories":"Research, Retirement","stInode":"be205f47-e7fa-4d89-aefd-3249aa11bea1"}]}
```

As you can see, the "Private Baking" category was not added to the content because there's no match with the passed in value and any Category Key/Variable (as per the screenshot above).

6. Edit the content in the backend and add "Private Banking" category. Save/Publish it.

<img width="1508" alt="screen shot 2017-10-03 at 1 52 28 pm" src="https://user-images.githubusercontent.com/1717110/31146044-a3731f64-a842-11e7-98cd-0b76d546daee.png">

7. Pull the content using REST API calls. As you can see, "Private Banking" is added to the list of categories associated to this content.

```
$ curl -v -u admin@dotcms.com:admin -XGET http://localhost:8080/api/content/query/+contentType:TestContent002
Note: Unnecessary use of -X or --request, GET is already inferred.
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
* Server auth using Basic with user 'admin@dotcms.com'
> GET /api/content/query/+contentType:TestContent002 HTTP/1.1
> Host: localhost:8080
> Authorization: Basic YWRtaW5AZG90Y21zLmNvbTphZG1pbg==
> User-Agent: curl/7.54.0
> Accept: */*
> 
< HTTP/1.1 200 OK
< Server: Apache-Coyote/1.1
< Set-Cookie: JSESSIONID=4943C389483892B2603EE3EFC04067F8; Path=/; HttpOnly
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET, HEAD, POST, PUT, DELETE, OPTIONS
< Access-Control-Allow-Headers: Authorization, Accept, Content-Type, Cookies
< Content-Type: text/plain
< Content-Length: 506
< Date: Tue, 03 Oct 2017 19:52:42 GMT
< 
* Connection #0 to host localhost left intact
{"contentlets":[{"owner":"dotcms.org.1","identifier":"f48d4c83-98aa-4157-9743-bb3936d15a09","modDate":"2017-10-03 13:52:24.0","languageId":1,"title":"Test 001 - Created through curl","inode":"aad5deae-2d95-4d25-af92-47c408044708","folder":"SYSTEM_FOLDER","sortOrder":0,"modUser":"dotcms.org.1","host":"SYSTEM_HOST","lastReview":"2017-10-03 13:52:24.0","__NAME__":"f48d4c83-98aa-4157-9743-bb3936d15a09","categories":"Private Banking, Research, Retirement","stInode":"be205f47-e7fa-4d89-aefd-3249aa11bea1"}]}
```

8. Update the content via REST API carl, by altering the Title only. **Re-use the Categories Names as the array returned on step 7**:

```
$ curl -v -u "admin@dotcms.com:admin" -XPUT http://localhost:8080/api/content/publish/1 -F 'json={"stName":"TestContent002","identifier":"f48d4c83-98aa-4157-9743-bb3936d15a09","title":"Test 001 - Updated through curl","categories":"Research, Retirement, Private Banking"};type=application/json'
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 8080 (#0)
* Server auth using Basic with user 'admin@dotcms.com'
> PUT /api/content/publish/1 HTTP/1.1
> Host: localhost:8080
> Authorization: Basic YWRtaW5AZG90Y21zLmNvbTphZG1pbg==
> User-Agent: curl/7.54.0
> Accept: */*
> Content-Length: 345
> Expect: 100-continue
> Content-Type: multipart/form-data; boundary=------------------------fc50878994be07cd
> 
< HTTP/1.1 100 Continue
< HTTP/1.1 200 OK
< Server: Apache-Coyote/1.1
< Set-Cookie: JSESSIONID=706ADA16BE59489A6D5AF62B4A687AFB; Path=/; HttpOnly
< Location: http://localhost:8080/api/content/inode/c6937e4c-db92-411d-a331-854289f28ce2
< inode: c6937e4c-db92-411d-a331-854289f28ce2
< identifier: f48d4c83-98aa-4157-9743-bb3936d15a09
< Access-Control-Allow-Origin: *
< Access-Control-Allow-Methods: GET, HEAD, POST, PUT, DELETE, OPTIONS
< Access-Control-Allow-Headers: Authorization, Accept, Content-Type, Cookies
< Content-Length: 0
< Date: Tue, 03 Oct 2017 19:55:01 GMT
< 
* Connection #0 to host localhost left intact
```

This is how the content was updated. See there's no "Private Banking" category assigned to it, meaning we cannot reuse the Category Names we got in the first place. Other categories remain in place because their Key/Variable value match their Category Names.

<img width="1507" alt="screen shot 2017-10-03 at 1 55 49 pm" src="https://user-images.githubusercontent.com/1717110/31146100-becb8710-a842-11e7-9408-c6de06983cd8.png">

## Context
Reported after making some tests as per this ticket https://dotcms.zendesk.com/agent/tickets/94450

## Your Environment

* dotCMS version used: 4.1.1
* Browser Name and version: Any Browser.
* Operating System and version: Any OS.
* Application Server and version: Any App server.
* Java Brand and version: Java 8u121.
* Database and version: Any DB.
* Application module (if apply):N/A

**Comment by mattyarbrough** [comment]
This is still an issue in 5.2.3

**Comment by zlymeda** [comment]
I was provided with hotfix for 5.1.6. Can the same hotfix be applied to master with exception for it to be configurable?
Thank you

**Comment by dsolistorres** [comment]
PR: #17915 
This PR adds a `allCategoriesInfo` parameter to Content REST API. When the parameter is used, all the categories information is returned.

**Comment by dsilvam** [comment]
Note to Doc: @john-thomas-dotcms 
Parameter `includeCategoriesExtraInfo` was added to the Content REST API. When the parameter is used, all the categories information is returned.
Example usage:  `/api/content/id/{ContentIdentifier}/includeCategoriesExtraInfo/true`

**Comment by dsilvam** [comment]
**Passed Internal QA**: categories extra info is included when the new parameter is provided via URI. 
URI: http://localhost:8080/api/content/id/76e07a19-6264-4d46-a6b1-922782e45c5b/includeCategoriesExtraInfo/true

Categories extra info:
```
"ageGender": [
                {
                    "inode": "5da7ef42-bfcc-42ee-87e7-59c5d93b1233",
                    "keywords": "",
                    "sortOrder": 0,
                    "description": null,
                    "categoryName": "Mens",
                    "key": "mens"
                },
                {
                    "inode": "1278044d-29db-4dee-a2ef-3eae7d7f793f",
                    "keywords": "",
                    "sortOrder": 0,
                    "description": null,
                    "categoryName": "Womens",
                    "key": "womens"
                }
            ]
```

**Comment by bryanboza** [comment]
Fixed, tested on release-5.2.6 // Postgres // FF
